File: profiling.py
import os
import subprocess
import sys
from datetime import datetime


# universal_domain = 1000000000
universal_domain = 10**9
# universal_domain = 1000000
# universal_iters = 5
universal_iters = 1


def runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters, random_range_size='false', min_range_size=0, max_range_size=0, percentage_point_queries=0):
    print('iter,strategy,distribution,num queries,domain size,leaf size,range size,range size min, range size avg, range size max,avg node length,# leaf nodes,max tree depth,mapping queries nodes,mapping insert ops,mapping transfer ops,mapping share ops,mapping merge ops,mapping insert time,mapping transfer time,mapping share time,mapping merge time,tree building time,mapping total time,additional tree time,mapping + tree building time,total time')

    for distribution in distributions:
        for strategy in strategies:
            for query_size in queries:
                for leaf_size in leaf_sizes:
                    for range_size in range_sizes:
                        for i in range(iters):
                            command = f"./profiling.out --pre_partitioning --distribution={distribution} --iter={i} --strategy={strategy} --queries={query_size} --key_domain_size={domain} --leaf_size={leaf_size} --range_size={range_size} --random_range_size={random_range_size} --min_range_size={min_range_size} --max_range_size={max_range_size} --percentage_point_queries={percentage_point_queries}"
                            os.system(command)


def runner_grouped(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters, random_range_size='false', min_range_size=0, max_range_size=0, percentage_point_queries=0):
    print('iter,strategy,distribution,num queries,domain size,leaf size,range size,range size min, range size avg, range size max,avg node length,# leaf nodes,max tree depth,mapping queries nodes,mapping insert ops,mapping transfer ops,mapping share ops,mapping merge ops,mapping insert time,mapping transfer time,mapping share time,mapping merge time,tree building time,mapping total time,additional tree time,mapping + tree building time,total time')

    for distribution in distributions:
        for strategy in strategies:
            for query_size in queries:
                for (leaf_size, range_size) in zip(leaf_sizes, range_sizes):
                    for i in range(iters):
                        command = f"./profiling.out --pre_partitioning --distribution={distribution} --iter={i} --strategy={strategy} --queries={query_size} --key_domain_size={domain} --leaf_size={leaf_size} --range_size={range_size} --random_range_size={random_range_size} --min_range_size={min_range_size} --max_range_size={max_range_size} --percentage_point_queries={percentage_point_queries}"
                        os.system(command)


def experiment_1a():
    print("############## 1 A #################")
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
    ]
    queries = [
        # 10,
        # 100,
        1000,
        10000,
        # 100000,
        # 1000000,
        # 10000000,
    ]
    leaf_sizes = [
        100000,
    ]
    range_sizes = [
        100000,
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_1b():
    print("############## 1 B #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        10,
        100,
        1000,
    ]
    leaf_sizes = [
        'max_range',
    ]
    range_sizes = [
        10000,
        100000,
        1000000,
        10000000,
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_2a():
    print("############## 2 A #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        int(10 ** 5),
    ]
    leaf_sizes = [
        int(universal_domain * 1 / 100),
    ]
    range_sizes = [
        int(universal_domain * 0.001 / 100),
        int(universal_domain * 0.01 / 100),
        int(universal_domain * 0.1 / 100),
        int(universal_domain * 1 / 100),
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_2b():
    print("############## 2 B #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        int(10 ** 6),
    ]
    leaf_sizes = [
        int(universal_domain * 1 / 100),
    ]
    range_sizes = [
        int(universal_domain * 0.001 / 100),
        int(universal_domain * 0.01 / 100),
        int(universal_domain * 0.1 / 100),
        int(universal_domain * 1 / 100),
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_2c():
    print("############## 2 C #################")
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
    ]
    queries = [
        100000,
    ]
    leaf_sizes = [
        int(universal_domain * (10 ** -5) / 100),
        int(universal_domain * (10 ** -4) / 100),
        int(universal_domain * (10 ** -3) / 100),
        int(universal_domain * (10 ** -2) / 100),
        int(universal_domain * (10 ** -1) / 100),
    ]
    range_sizes = [
        int(universal_domain * (10 ** -5) / 100),
        int(universal_domain * (10 ** -4) / 100),
        int(universal_domain * (10 ** -3) / 100),
        int(universal_domain * (10 ** -2) / 100),
        int(universal_domain * (10 ** -1) / 100),
    ]
    distributions = ['normal']
    domain = universal_domain

    runner_grouped(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_3a():
    print("############## 3 A #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        100000,
    ]
    leaf_sizes = [
        int(universal_domain * (10 ** -4) / 100),
        int(universal_domain * (10 ** -3) / 100),
        int(universal_domain * (10 ** -2) / 100),
        int(universal_domain * (10 ** -1) / 100),
        int(universal_domain * (10 ** 0) / 100),
    ]
    range_sizes = [
        int(universal_domain * (10 ** -3) / 100),
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_3b():
    print("############## 3 B #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        100000,
    ]
    leaf_sizes = [
        int(universal_domain * (10 ** -4) / 100),
        int(universal_domain * (10 ** -3) / 100),
        int(universal_domain * (10 ** -2) / 100),
        int(universal_domain * (10 ** -1) / 100),
        int(universal_domain * (10 ** 0) / 100),
    ]
    range_sizes = [
        int(universal_domain * (10 ** -4) / 100),
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_3c():
    print("############## 3 C #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        100000,
    ]
    leaf_sizes = [
        int(universal_domain * (10 ** -4) / 100),
        int(universal_domain * (10 ** -3) / 100),
        int(universal_domain * (10 ** -1) / 100),
        int(universal_domain * (10 ** 0) / 100),
        'max_range',
    ]
    range_sizes = [
        0,
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(
        distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters,
        random_range_size='true',
        # min_range_size=int(universal_domain * (10 ** -4) / 100),
        # max_range_size=int(universal_domain * (10 ** -3) / 100),
    )


def experiment_3d():
    print("############## 3 D #################")
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
    ]
    queries = [
        100000,
    ]
    leaf_sizes = [
        10**3,
        # 10**4,
        # 10**5,
        # 10**6,
        # 10**7,
        # 10**8,
        # 10**9,
    ]
    range_sizes = [
        10**5,
    ]
    distributions = ['normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_4():
    print("############## 4 #################")
    iters = universal_iters

    strategies = [
        'additional',
        'lazy',
        'eager',
    ]
    queries = [
        1000,
        10000,
        100000,
    ]
    leaf_sizes = [
        int(universal_domain * 1 / 100),
    ]
    range_sizes = [
        int(universal_domain * (10 ** -4) / 100),
    ]
    distributions = ['zipf', 'normal']
    domain = universal_domain

    runner(distributions, domain, strategies, queries, leaf_sizes, range_sizes, iters)


def experiment_6():
    point_queries = [
        50,
        80,
        90,
        99,
    ]
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
        # 'qat',
    ]
    queries = [
        10000,
    ]
    leaf_sizes = [
        "max_range",
    ]
    range_sizes = [
        100,
        1000,
        10000,
        100000,
    ]
    distribution = 'normal'
    domain = universal_domain

    for i in point_queries:
        print(f"############## 6 {i} #################")
        runner(
            distribution, domain, strategies, queries,
            leaf_sizes, range_sizes, iters,
            percentage_point_queries=i)


def experiment_61():
    print("############## 6.1 A #################")
    iters = universal_iters

    strategies = [
        # 'additional',
        # 'lazy',
        # 'eager',
        'qat',
    ]
    queries = [
        10000,
    ]
    leaf_sizes = [
        "max_range",
    ]
    range_sizes = [
        100,
        1000,
        10000,
        100000,
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters, percentage_point_queries=50)


def experiment_62():
    print("############## 6.2 A #################")
    iters = universal_iters

    strategies = [
        # 'additional',
        # 'lazy',
        # 'eager',
        'qat',
    ]
    queries = [
        10000,
    ]
    leaf_sizes = [
        "max_range",
    ]
    range_sizes = [
        100,
        1000,
        10000,
        100000,
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters, percentage_point_queries=80)


def experiment_63():
    print("############## 6.3 A #################")
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
        # 'qat',
    ]
    queries = [
        10000,
    ]
    leaf_sizes = [
        "max_range",
    ]
    range_sizes = [
        100,
        1000,
        10000,
        100000,
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters, percentage_point_queries=90)


def experiment_64():
    print("############## 6.4 A #################")
    iters = universal_iters

    strategies = [
        'additional',
        # 'lazy',
        # 'eager',
        # 'qat',
    ]
    queries = [
        10000,
    ]
    leaf_sizes = [
        "max_range",
    ]
    range_sizes = [
        100,
        1000,
        10000,
        100000,
    ]
    distribution = 'normal'
    domain = universal_domain

    runner(distribution, domain, strategies, queries, leaf_sizes, range_sizes, iters, percentage_point_queries=99)


if __name__ == '__main__':
    experiment_1a()
    # experiment_1b()
    # experiment_2a()
    # experiment_2b()
    # experiment_2c()
    # experiment_3a()
    # experiment_3b()
    # experiment_3c()
    # experiment_3d()
    # experiment_4()
    # experiment_6()
    # experiment_61()
    # experiment_62()
    # experiment_63()
    # experiment_64()

File: profiling.cpp
#include <gflags/gflags.h>

#include "../interval-base-tree/src/newtree.h"

#include "result.h"
#include "utils.h"

DEFINE_int64(queries, 100, "Number of queries");
DEFINE_int64(key_domain_size, 1000000, "Key domain size");
DEFINE_string(leaf_size, "100000", "Leaf size");
DEFINE_int64(range_size, 100000, "Range of queries");
DEFINE_bool(random_range_size, false, "Use random range of queries");
DEFINE_int64(percentage_point_queries, 0, "% of Pointe Queries");
DEFINE_string(strategy, "raw", "Strategy");
DEFINE_string(distribution, "normal", "Random Distribution");
DEFINE_int64(iter, 0, "Define the iteration number");
DEFINE_int64(seed, 100, "Random Seed");


void printTimes(T * queriesMeta, Tree <Traits <T> > * & tree, double total_time, double mapping_time = 0) {
    double tbt, mt, t2t, ttt;
    tbt = total_time - tree->qMap->elapsedTime();
    T * leafsData = tree->getLeafsData();

    cout << FLAGS_iter << ",";
    cout << FLAGS_strategy << "," << FLAGS_distribution << "," << FLAGS_queries << ",";
    cout << FLAGS_key_domain_size << "," << FLAGS_leaf_size << "," << FLAGS_range_size << ",";
    cout << queriesMeta[0] << "," << queriesMeta[1] << "," << queriesMeta[2] << ",";
    cout << leafsData[0] << "," << leafsData[3] << "," << leafsData[4] << ",";
    cout << tree->qMap->csv() << ",";
    if (mapping_time > 0) {
        t2t = mapping_time;
        mt = 0;
        ttt = tbt + mapping_time;
    } else {
        t2t = 0;
        mt = tree->qMap->elapsedTime();
        ttt = total_time;
    }
    cout << tbt << "," << mt << "," << t2t << "," << ttt;

    cout << endl;
}

void additional(vector <Tinterval> & queries, T leaf_size, T * queriesMeta) {
    auto start_time = std::chrono::system_clock::now();
    QMapExtra <Traits <T>> * qMap = new QMapExtra <Traits <T>>();
    Tree <Traits <T> > * tree = new Tree <Traits <T> >(leaf_size, qMap);

    for (int i = 0; i < queries.size(); i += 1) {
        tree->insert(queries[i]);
    }

    auto end_time = std::chrono::system_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    double total_time = elapsed_seconds.count();

    start_time = std::chrono::system_clock::now();
    LeafTree<Traits <T>> leaftree;
    vector<Node<T> *> leafs;
    tree->root->getLeafs(leafs);
    random_shuffle(leafs.begin(), leafs.end());

    for (int i = 0; i < leafs.size(); i++) {
        leaftree.insert(leafs[i]->interval);
    }

    for (int i = 0; i < queries.size(); i += 1) {
        leaftree.assign(&queries[i]);
    }
    tree->qMap->indexed = leaftree.numIndexedQueries();

    end_time = std::chrono::system_clock::now();
    elapsed_seconds = end_time - start_time;
    double mapping_time = elapsed_seconds.count();

    printTimes(queriesMeta, tree, total_time, mapping_time);
}

void eager(vector <Tinterval> & queries, T leaf_size, T * queriesMeta) {
    auto start_time = std::chrono::system_clock::now();
    QMapEager <Traits <T>> * qMap = new QMapEager <Traits <T>>();
    Tree <Traits <T> > * tree = new Tree <Traits <T> >(leaf_size, qMap);

    for (int i = 0; i < queries.size(); i += 1) {
        tree->insert(queries[i]);
    }

    auto end_time = std::chrono::system_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    double total_time = elapsed_seconds.count();

    printTimes(queriesMeta, tree, total_time);
}

void lazy(vector <Tinterval> & queries, T leaf_size, T * queriesMeta) {
    auto start_time = std::chrono::system_clock::now();
    QMapLazy <Traits <T>> * qMap = new QMapLazy <Traits <T>>();
    Tree <Traits <T> > * tree = new Tree <Traits <T> >(leaf_size, qMap);

    for (int i = 0; i < queries.size(); i += 1) {
        tree->insert(queries[i]);
    }

    auto end_time = std::chrono::system_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    double total_time = elapsed_seconds.count();

    printTimes(queriesMeta, tree, total_time);
}

int main(int argc, char** argv) {
    gflags::SetUsageMessage("Profiling");
    gflags::SetVersionString("1.0.0");
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    if (FLAGS_range_size > FLAGS_key_domain_size) {
        cout << "Range Size should be lower that domain" << endl;
        return 0;
    }

    srand (FLAGS_seed);
    vector <Tinterval> queries;
    T leaf_size = 0;

    if (FLAGS_distribution == "zipf") {
        queries = create_queries_zipf(FLAGS_queries, FLAGS_key_domain_size, FLAGS_range_size, FLAGS_random_range_size, FLAGS_percentage_point_queries);
    } else {
        queries = create_queries(FLAGS_queries, FLAGS_key_domain_size, FLAGS_range_size, FLAGS_random_range_size, FLAGS_percentage_point_queries);
    }

    T * queriesMeta = getQueriesMeta(queries);
    if (is_number(FLAGS_leaf_size)) {
        leaf_size = atol(FLAGS_leaf_size.c_str());
    } else if (FLAGS_leaf_size == "max_range") {
        leaf_size = queriesMeta[2];
    }

    if (FLAGS_strategy == "lazy") {
        lazy(queries, leaf_size, queriesMeta);
    } else if (FLAGS_strategy == "eager") {
        eager(queries, leaf_size, queriesMeta);
    } else if (FLAGS_strategy == "additional") {
        additional(queries, leaf_size, queriesMeta);
    }
}
